packages/keyring-controller/src/KeyringController.ts

```diff
@@ -1299,6 +1309,69 @@ export class KeyringController extends BaseController<
     return seedWords;
   }
 
+  /**
+   * Select a keyring and execute the given function with
+   * the selected keyring, as a mutually exclusive atomic
+   * operation.
+   *
+   * The method automatically persists changes at the end of the
+   * function execution, or rolls back the changes if an error
+   * is thrown.
+   *
+   * @param selector - Keyring selector object.
+   * @param fn - Function to execute with the selected keyring.
+   * @param options - Additional options.
+   * @param options.createIfMissing - Whether to create a new keyring if the selected one is missing.
+   * @param options.createWithData - Optional data to use when creating a new keyring.
+   * @returns Promise resolving to the result of the function execution.
+   */
+  async withKeyring<K extends EthKeyring<Json> = EthKeyring<Json>, T = void>(
+    selector: KeyringSelector,
+    fn: (keyring: K) => Promise<T>,
+    options:
+      | { createIfMissing?: false }
+      | { createIfMissing: true; createWithData?: unknown } = {
+      createIfMissing: false,
+    },
+  ): Promise<T> {
+    return this.#persistOrRollback(async () => {
+      let keyring: K | undefined;
+
+      if ('address' in selector) {
+        keyring = (await this.getKeyringForAccount(
+          normalize(selector.address) as Hex,
+        )) as K | undefined;
+      } else {
+        keyring = this.getKeyringsByType(selector.type)[selector.index || 0] as
+          | K
+          | undefined;
+
+        if (!keyring && options.createIfMissing) {
```

Comment:
Hmm. Do we need to include this `createIfMissing` functionality? I know that the QR keyring does this, but in general this pattern of creating a keyring if it's missing is I think something we're aiming to move away from. Embedding that pattern here might encourage its use.

Perhaps we could continue to use `getOrAddQrKeyring()` for creating the keyring, and leave creation out of this method.

Reply:
As discussed during Office Hours, I overload the function and deprecated the alternative with `options`: [fdc0c8d](https://github.com/MetaMask/core/pull/4197/commits/fdc0c8dd48752a69e027cb0a7a8089d1133bf1d6)